vmdk_plugin/plugin.go

```diff
@@ -117,19 +120,70 @@ func (d *vmdkDriver) mountVolume(name string) (string, error) {
 		return mountpoint, err
 	}
 
+	if d.useMockEsx {
+		return mountpoint, fmt.Errorf("No device to mount.")
+	}
+
+	skipInotify := false
+
+	watcher, err := inotify.NewWatcher()
+
+	if err != nil {
+		log.WithFields(
+			log.Fields{"name": name, "dir": mountpoint},
+		).Error("Failed to create watcher, skip inotify ")
+		skipInotify = true
+	} else {
+		err = watcher.Watch(watchPath)
+		if err != nil {
+			log.WithFields(
+				log.Fields{"name": name, "dir": mountpoint},
+			).Error("Failed to watch /dev, skip inotify ")
+			skipInotify = true
+		}
+	}
+
 	// Have ESX attach the disk
 	dev, err := d.ops.Attach(name, nil)
 	if err != nil {
 		return mountpoint, err
 	}
 
-	time.Sleep(3 * time.Second)
+	device, err := fs.GetDevicePath(dev)
+	if err != nil {
+		return mountpoint, err
+	}
 
-	if d.useMockEsx {
-		return mountpoint, fs.Mount(mountpoint, nil, "ext4")
+	if skipInotify {
+		time.Sleep(sleepBeforeMount)
+		return mountpoint, fs.Mount(mountpoint, "ext2", device)
+	}
+loop:
+	for {
+		select {
+		case ev := <-watcher.Event:
+			log.Debug("event: ", ev)
+			if ev.Name == device {
+				// Log when the device is discovered
+				log.WithFields(
+					log.Fields{"name": name, "event": ev},
+				).Info("Attach complete ")
+				break loop
+			}
+		case err := <-watcher.Error:
+			log.WithFields(
+				log.Fields{"name": name, "device": device, "error": err},
+			).Error("Hit error during watch ")
+			break loop
+		case <-time.After(devWaitTimeout):
+			log.WithFields(
+				log.Fields{"name": name, "timeout": devWaitTimeout, "device": device},
+			).Error("Reached timeout while waiting for device ")
```

Comment:
What if the device is attached now by the OS, the device is attached and does docker detach if there is an error?

Should have a test to force both the error and this timeout path (return error from here even though the disk is attached) and figure that the disk isn't stuck to the VM and unusable.